### PR TITLE
Implement `where T: class` and `where T: struct` for overrides and implementations

### PR DESCRIPTION
**Pull request by YairHalberstadt**
https://github.com/dotnet/roslyn/issues/34798#issuecomment-483414230

### CODE REVIEW

**Comment by YairHalberstadt** [comment]
Still to do:
Fix broken tests.
_declaredConstraints in SourceOrdinaryMethodSymbol can be a local, not a field.

**Comment by YairHalberstadt** [comment]
@AlekseyTs 
I think this is ready for review.

**Comment by AlekseyTs** [comment]
Thank you, @YairHalberstadt, I will take it from here and will make any necessary adjustments.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs [hunk 1]
```diff
@@ -384,20 +429,48 @@ private void MethodChecks(MethodDeclarationSyntax syntax, Binder withTypeParamsB
                 }
             }
 
+            if (declaredConstraints is { })
```
Comment (jmarolf) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs]: I believe `is null` or `is object` would be clearer

**Comment by AlekseyTs** [comment]
Closing in favor of https://github.com/dotnet/roslyn/pull/35048

**Pull request closed without merging**